### PR TITLE
[WIP] WebTorrent Support

### PR DESCRIPTION
**Pull request by yrliou**
TODO:
- Automation test
- Add webtorrent and bittorrent-tracker forks into vendor and use them instead
- Investigate DCHECK error caused by sockets.tcp.connect

What's included in this PR:
- WebTorrent support of magnet links

What's not included in this PR:
- WebTorrent support of .torrent links
- Support of trackless torrents using bittorrent-dht/client
- l10n

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Needed or QA/No-QA-Needed) to include the closed issue in milestone 

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: patches/chrome-browser-extensions-component_extensions_whitelist-whitelist.cc.patch [hunk 1]
```diff
@@ -18,11 +18,12 @@ index dfc12043f6c68fd0a258ae54cb2dd8e12f9117a0..2e56d1e21fbc2a6ba1149be0de102a80
    };
  
    for (size_t i = 0; i < arraysize(kAllowed); ++i) {
-@@ -60,6 +62,7 @@ bool IsComponentExtensionWhitelisted(const std::string& extension_id) {
+@@ -60,6 +62,8 @@ bool IsComponentExtensionWhitelisted(const std::string& extension_id) {
  bool IsComponentExtensionWhitelisted(int manifest_resource_id) {
    switch (manifest_resource_id) {
      // Please keep the list in alphabetical order.
 +    case IDR_BRAVE_EXTENSON:
++    case IDR_BRAVE_WEBTORRENT:
```
Comment (yrliou) [inline patches/chrome-browser-extensions-component_extensions_whitelist-whitelist.cc.patch]: I'll adapt @bridiver's work in sync WIP on moving it to chromium_src

File: patches/net-url_request-url_request_context_builder.cc.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/net/url_request/url_request_context_builder.cc b/net/url_request/url_request_context_builder.cc
+index 800515a58cfa8546288d961fc9d14ae4676fcb5f..c75eec623aa0aff28ed8fd0607af082bc58de768 100644
+--- a/net/url_request/url_request_context_builder.cc
++++ b/net/url_request/url_request_context_builder.cc
+@@ -621,6 +621,8 @@ std::unique_ptr<URLRequestContext> URLRequestContextBuilder::Build() {
+   }
+ #endif  // !BUILDFLAG(DISABLE_FTP_SUPPORT)
+ 
++  job_factory->SetProtocolHandler("magnet", std::make_unique<MagnetProtocolHandler>());
```
Comment (bridiver) [inline patches/net-url_request-url_request_context_builder.cc.patch]: this should not be added here. See ProfileIOData::AddProtocolHandlersToBuilder and ProfileIOData::InstallProtocolHandlers

File: script/build-brave-webtorrent.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import argparse
```
Comment (bridiver) [inline script/build-brave-webtorrent.py]: we already have a `transpile_web_ui` action and it should be adapted if any changes are needed to handle this

File: .gitignore [hunk 4]
```diff
@@ -2,6 +2,7 @@
 .tags*
 /.idea/
 /browser/resources/brave_extension/
+/browser/resources/brave_webtorrent/*.bundle.js
```
Comment (bridiver) [inline .gitignore]: we shouldn't be writing back into the source directory. Please see the wip sync branch

File: browser/resources/brave_webtorrent/BUILD.gn [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import("//tools/grit/grit_rule.gni")
```
Comment (bridiver) [inline browser/resources/brave_webtorrent/BUILD.gn]: I think this should all move into `components/webtorrent` and resources should be kept inside the component like sync

File: browser/resources/brave_extension.grd [hunk 6]
```diff
@@ -20,6 +20,11 @@
       <include name="IDR_BRAVE_EXTENSON_EN_US_MESSAGES_JSON" file="brave_extension/_locales/en_US/messages.json" type="BINDATA" />
       <include name="IDR_BRAVE_EXTENSON_BRAVE_SHIELDS_HTML" file="brave_extension/braveShieldsPanel.html" type="BINDATA" />
       <include name="IDR_BRAVE_EXTENSON_BRAVELIZER_JS" file="brave_extension/bravelizer.css" type="BINDATA" />
+
+      <!-- Brave Webtorrent -->
```
Comment (bridiver) [inline browser/resources/brave_extension.grd]: see https://github.com/brave/brave-core/pull/226/commits/88818ab52f63221c683f64c8787f0acce15ee087

File: script/build-brave-webtorrent.py (hunk 3 above)
Comment (yrliou) [reply script/build-brave-webtorrent.py]: addressed by https://github.com/brave/brave-core/pull/353/commits/32c0fe277d781f756ab5581b213b01aef1986764

File: .gitignore (hunk 4 above)
Comment (yrliou) [reply .gitignore]: It seems we couldn't access these *.bundle.js files from the extension if it's in gen, so I'm still generating webpack output files into src, they're now under `components/brave_webtorrent/entension/out`.

File: browser/resources/brave_webtorrent/BUILD.gn (hunk 5 above)
Comment (yrliou) [reply browser/resources/brave_webtorrent/BUILD.gn]: addressed by 32c0fe2

File: browser/resources/brave_extension.grd (hunk 6 above)
Comment (yrliou) [reply browser/resources/brave_extension.grd]: addressed by 32c0fe2

**Comment by yrliou** [comment]
Oops... sorry, I accidentally delete this remote branch, I'll open a new PR.

**Pull request closed without merging**